config/dynamic.c

```diff
@@ -16,11 +16,527 @@
+		 * The config component accepts only one connection at a time.
+		 */
+		if (FD_ISSET(dy_conf->sock_fd, &fds))
+			handle_client(dy_conf->sock_fd, dy_conf);
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"dyn_cfg: the Dynamic Config block at lcore = %u is exiting\n", lcore);
+
+	cleanup_dy(dy_conf);
+
+	return ret;
+}
+
+struct dynamic_config *
+alloc_dy_conf(void)
+{
+	return rte_calloc("dynamic_config",
+		1, sizeof(struct dynamic_config), 0);
+}
+
+void
+set_dyc_timeout(unsigned int sec,
+	unsigned int usec, struct dynamic_config *dy_conf)
+{
+	dy_conf->rcv_time_out.tv_sec = sec;
+	dy_conf->rcv_time_out.tv_usec = usec;
+}
 
 int
-run_dynamic_config(__attribute__((unused)) const struct dynamic_config *dy_conf)
+run_dynamic_config(const char *server_path, struct dynamic_config *dy_conf)
 {
-	/* TODO Initialize and run Dynamic Config functional block. */
+	int ret;
+	struct sockaddr_un server_addr;
+
+	if (server_path == NULL || dy_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	dy_conf->sock_fd = -1;
+
+	dy_conf->server_path = rte_malloc(
+		"server_path", strlen(server_path) + 1, 0);
+	if (dy_conf->server_path == NULL) {
+		ret = -1;
+		goto out;
+	}
+	strcpy(dy_conf->server_path, server_path);
+
+	/*
+	 * Remove any old socket and create an unnamed socket for the server.
+	 */
+	ret = unlink(dy_conf->server_path);
+	if (ret != 0 && errno != ENOENT) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to unlink(%s) - (%s)\n",
+			dy_conf->server_path, strerror(errno));
+		ret = -1;
+		goto free_path;
+	}
+
+	/* Init the server socket. */
+    	dy_conf->sock_fd = socket(AF_UNIX, SOCK_STREAM, 0);
+	if (dy_conf->sock_fd < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to initialize the server socket - (%s)\n",
+			strerror(errno));
+		ret = -1;
+		goto free_path;
+	}
+
+	/* Name the socket. */
+    	memset(&server_addr, 0, sizeof(server_addr));
+    	server_addr.sun_family = AF_UNIX;
+
+	if (sizeof(server_addr.sun_path) <= strlen(dy_conf->server_path)) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: the server path (%s) exceeds the length limit %lu\n",
+			dy_conf->server_path, sizeof(server_addr.sun_path));
+		ret = -1;
+		goto free_sock;
+	}
+
+    	strcpy(server_addr.sun_path, dy_conf->server_path);
+
+    	ret = bind(dy_conf->sock_fd,
+		(struct sockaddr *)&server_addr, sizeof(server_addr));
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to bind the server socket - (%s)\n",
+			strerror(errno));
+		ret = -1;
+		goto free_sock;
+	}
+
+	/*
+	 * The Dynamic config component listens to a Unix socket
+	 * for request from the local host.
+	 */
+    	ret = listen(dy_conf->sock_fd, 10);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to listen on the server socket - (%s)\n",
+			strerror(errno));
+		ret = -1;
+		goto free_sock;
+	}
+
+	ret = launch_at_stage3("dynamic_conf",
+		dyn_cfg_proc, dy_conf, dy_conf->lcore_id);
+	if (ret < 0)
+		goto free_sock;
+
 	return 0;
+
+free_sock:
+	ret = close(dy_conf->sock_fd);
```

Comment:
We cannot change `ret` here. Consider this case: `close()` was successful, the function is going to return success (i.e. zero) although there was an error.